src/ripple/app/main/GRPCServer.h

```diff
@@ -0,0 +1,349 @@
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_CORE_GRPCSERVER_H_INCLUDED
+#define RIPPLE_CORE_GRPCSERVER_H_INCLUDED
+
+#include <ripple/rpc/Role.h>
+#include <ripple/rpc/Context.h>
+#include <ripple/rpc/GRPCHandlers.h>
+#include <ripple/rpc/impl/RPCHelpers.h>
+#include <ripple/rpc/impl/Handler.h>
+#include <ripple/rpc/impl/Tuning.h>
+#include <ripple/net/InfoSub.h>
+#include <ripple/protocol/ErrorCodes.h>
+#include <ripple/core/JobQueue.h>
+#include <ripple/resource/Charge.h>
+
+#include <grpcpp/grpcpp.h>
+#include "rpc/v1/xrp_ledger.grpc.pb.h"
+
+namespace ripple {
+
+//Interface that CallData implements
+class Processor
+{
+    public:
+
+    virtual ~Processor() {}
+
+    // process a request that has arrived. Can only be called once per instance
+    virtual void process() = 0;
+
+    //store an iterator to this object
+    //all Processor objects are stored in a std::list, and the iterator points
+    //to that objects position in the list. When object finishes processing a 
+    //request, the iterator is used to delete the object from the list
+    virtual void set_iter(
+            std::list<std::shared_ptr<Processor>>::iterator const& it) = 0;
+
+    //get iterator to this object. see above comment
+    virtual std::list<std::shared_ptr<Processor>>::iterator get_iter() = 0;
+
+    //abort processing this request. called when server shutsdown
+    virtual void abort() = 0;
+
+    //create a new instance of this CallData object, with the same type
+    //(same template parameters) as original. This is called when a CallData
+    //object starts processing a request. Creating a new instance allows the 
+    //server to handle additional requests while the first is being processed
+    virtual std::shared_ptr<Processor> clone() = 0;
+
+    //true if this object has finished processing the request. Object will be
+    //deleted once this function returns true
+    virtual bool isFinished() = 0;
+};
+
+namespace {
+
+//helper function. strips port from endpoint string
+std::string getEndpoint(std::string const& peer)
```

Comment:
this method is not ipv6 friendly. I'm not sure what format these peer strings have, but it seems like mainly they are endpoint strings with an optional scheme (`dns://`) ? See if `parseURL` will work for you (I think it will only parse if a scheme is present..not sure) or maybe just strip the scheme yourself by searching for `//` and then use `Endpoint::from_string` to do the rest.

Reply:
In my testing, the peer strings look like `ipv4:127.0.0.1:49808`. `parseUrl` is not working, and `Endpoint::from_string` does not work either. If I strip `ipv4` off of the front, I can then pass the string to `Endpoint::from_string`, which is actually what I'm doing already (if you look at where this function is used). I'm not sure the best way to do this; the documentation for the `peer()` method says a string is returned in uri format.

Reply:
I think what I have here will work, though correct me if I'm wrong. Apparently, the string is always of the form `protocol:ip:port`. See this PR for reference (you sent this to me actually): [https://github.com/grpc/grpc/pull/14667](https://github.com/grpc/grpc/pull/14667)